hera_cal/lstbin_simple.py

```diff
@@ -0,0 +1,1677 @@
+    out_data, out_flags, out_std, out_nsamples
+        The reduced data, flags, standard deviation (across days) and nsamples. 
+        Shape ``(nbl, nlst_bins, nfreq, npol)``.
+    """
+    nlst_bins = len(data)
+    (_, nbl, nfreq, npol) = data[0].shape
+
+    for d, f, n in zip(data, flags, nsamples):
+        assert d.shape == f.shape == n.shape
+
+    # Do this just so that we can save memory if the call to this function already
+    # has allocated memory.
+    if out_data is None:
+        out_data = np.zeros((nbl, nlst_bins, nfreq, npol), dtype=complex)
+    if out_flags is None:
+        out_flags = np.zeros(out_data.shape, dtype=bool)
+    if out_std is None:
+        out_std = np.ones(out_data.shape, dtype=complex)
+    if out_nsamples is None:
+        out_nsamples = np.zeros(out_data.shape, dtype=float)
+
+    assert out_data.shape == out_flags.shape == out_std.shape == out_nsamples.shape
+    assert out_data.shape == (nbl, nlst_bins, nfreq, npol)
+
+    for lstbin, (d,n,f) in enumerate(zip(data, nsamples, flags)):
+        logger.info(f"Computing LST bin {lstbin+1} / {nlst_bins}")
+        
+        # TODO: check that this doesn't make yet another copy...
+        # This is just the data in this particular lst-bin.
+        if d.size:
+            (
+                out_data[:, lstbin], 
+                out_flags[:, lstbin], 
+                out_std[:, lstbin], 
+                out_nsamples[:, lstbin]
+            ) = lst_average(
+                d, n, f, mutable=mutable, 
+                sigma_clip_thresh=sigma_clip_thresh, 
+                sigma_clip_min_N=sigma_clip_min_N
+            )
+        else:
+            out_data[:, lstbin] = 1.0
+            out_flags[:, lstbin] = True
+            out_std[:, lstbin] = 1.0
+            out_nsamples[:, lstbin] = 0.0
+
+        
+    return out_data, out_flags, out_std, out_nsamples
+
+def _allocate_dnf(shape: tuple[int], d=0.0, f=0, n=0):
+    data = np.full(shape, d, dtype=complex)
+    flags = np.full(shape, f, dtype=bool)
+    nsamples = np.full(shape, n, dtype=float)
+    return data, flags, nsamples
+
+@profile
+def lst_average(
+    data: np.ndarray, nsamples: np.ndarray, flags: np.ndarray, 
+    flag_thresh: float = 0.7, median: bool = False,
+    mutable: bool = False,
+    sigma_clip_thresh: float = 0.0,
+    sigma_clip_min_N: int=4,
+) -> tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray, np.ndarray]:
+    """
+    Compute statistics of a set of data over its first axis.
+
+    The idea here is that the data's first axis is "nights", and that each night is
+    at the same LST. However, this function is agnostic to the meaning of the first
+    axis. It just computes the mean, std, and nsamples over the first axis.
+
+    This function is meant to be used on a single element of a list returned by
+    :func:`simple_lst_bin`.
+
+    Parameters
+    ----------
+    data
+        The data to compute the statistics over. Shape ``(ntimes, nbl, nfreq, npol)``.
```

Comment:
shouldn't this be nominally "(nnights, ntimes, nbl, nfreq, npol)"? or am I confused?